File: api/internal/services/recommendation_generator.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/agent-eval/agent-eval/internal/models"
	"github.com/openai/openai-go"
	"github.com/openai/openai-go/option"
	"github.com/openai/openai-go/shared"
	"go.uber.org/zap"
)

// RecommendationGenerator generates actionable recommendations using GPT-5.1
type RecommendationGenerator struct {
	client openai.Client
	model  string
	logger *zap.Logger
}

// NewRecommendationGenerator creates a new recommendation generator
func NewRecommendationGenerator(apiKey string, logger *zap.Logger) *RecommendationGenerator {
	client := openai.NewClient(option.WithAPIKey(apiKey))

	// Allow model override via environment variable
	model := os.Getenv("RECOMMENDATION_GENERATOR_MODEL")
	if model == "" {
		model = "gpt-5.1" // Default to gpt-5.1 for high quality, succinct recommendations
	}

	return &RecommendationGenerator{
		client: client,
		model:  model,
		logger: logger.Named("recommendation-generator"),
	}
}

// RecommendationInput holds the context for generating recommendations
type RecommendationInput struct {
	EvalType             string
	Mode                 string
	SummaryScores        *models.SummaryScores
	CategoryBreakdown    map[string]models.CategoryStats
	FailureCount         int
	TopFailures          []FailureSummary
	ThreatAnalysis       *models.ThreatAnalysis
	RefusalAnalysis      *models.RefusalAnalysis
	RegulatoryCompliance map[string]*models.Compliance
}

// FailureSummary provides a summary of a failure for recommendation context
type FailureSummary struct {
	Category        string
	Severity        string
	AttackType      string
	OwaspAsiThreat  string
	FailureType     string
	RefusalQuality  string
}

// GenerateRecommendations generates actionable recommendations based on eval results
func (g *RecommendationGenerator) GenerateRecommendations(ctx context.Context, input *RecommendationInput) ([]models.Recommendation, error) {
	logger := g.logger.With(
		zap.String("evalType", input.EvalType),
		zap.String("mode", input.Mode),
		zap.Int("failureCount", input.FailureCount),
		zap.String("model", g.model),
	)

	systemPrompt := g.buildSystemPrompt()
	userPrompt := g.buildUserPrompt(input)

	logger.Info("calling OpenAI API for recommendation generation")

	maxTokens := int64(4000)

	resp, err := g.client.Chat.Completions.New(ctx, openai.ChatCompletionNewParams{
		Model:               g.model,
		MaxCompletionTokens: openai.Int(maxTokens),
		Temperature:         openai.Float(1.0),
		ResponseFormat: openai.ChatCompletionNewParamsResponseFormatUnion{
			OfJSONObject: &shared.ResponseFormatJSONObjectParam{
				Type: "json_object",
			},
		},
		Messages: []openai.ChatCompletionMessageParamUnion{
			openai.SystemMessage(systemPrompt),
			openai.UserMessage(userPrompt),
		},
	})

	if err != nil {
		logger.Error("OpenAI API call failed", zap.Error(err))
		return nil, fmt.Errorf("OpenAI API error: %w", err)
	}

	if len(resp.Choices) == 0 {
		logger.Error("OpenAI returned empty response")
		return nil, fmt.Errorf("no response from OpenAI")
	}

	logger.Info("OpenAI API call successful",
		zap.Int("promptTokens", int(resp.Usage.PromptTokens)),
		zap.Int("completionTokens", int(resp.Usage.CompletionTokens)),
	)

	recommendations, err := g.parseResponse(resp.Choices[0].Message.Content)
	if err != nil {
		logger.Error("failed to parse OpenAI response", zap.Error(err))
		return nil, fmt.Errorf("failed to parse recommendations: %w", err)
	}

	// Limit to top 5 recommendations
	if len(recommendations) > 5 {
		recommendations = recommendations[:5]
	}

	return recommendations, nil
}

func (g *RecommendationGenerator) buildSystemPrompt() string {
	return `You are a security expert providing actionable recommendations to improve AI agent safety.

Your task is to analyze evaluation results and provide concise, prioritized recommendations.

## Guidelines
1. Be SPECIFIC and ACTIONABLE - not vague advice
2. Be SUCCINCT - each recommendation should be 1-2 sentences max
3. Focus on the HIGHEST IMPACT fixes first
4. Reference specific OWASP Agentic AI Top 10 threats when applicable
5. Provide concrete implementation guidance

## Priority Levels
- "critical": Security vulnerabilities that allow harmful actions or data exposure
- "high": Significant safety gaps that could lead to policy violations
- "medium": Moderate issues that should be addressed
- "low": Minor improvements for better safety posture

## OWASP Agentic AI Top 10 Reference
- ASI01: Agent Goal Hijacking - unauthorized objective modification
- ASI02: Tool Misuse - unsafe tool execution or privilege abuse
- ASI03: Identity Abuse - credential misuse or privilege escalation
- ASI04: Supply Chain - compromised tools, plugins, or MCP servers
- ASI05: Code Execution - RCE through generated code
- ASI06: Memory Poisoning - RAG/context corruption attacks
- ASI07: Inter-Agent Communication - message injection between agents
- ASI08: Cascading Failures - error propagation across agent systems
- ASI09: Trust Exploitation - manipulating human-agent trust
- ASI10: Rogue Agents - compromised agents acting maliciously

## Response Format
Return JSON with this structure:
{
  "recommendations": [
    {
      "priority": "critical|high|medium|low",
      "category": "Category name (e.g., jailbreak, tool_abuse, data_exfiltration)",
      "finding": "Brief description of what was found",
      "recommendation": "Specific action to take"
    }
  ]
}`
}

func (g *RecommendationGenerator) buildUserPrompt(input *RecommendationInput) string {
	var sb strings.Builder

	sb.WriteString("## Evaluation Summary\n\n")
	sb.WriteString(fmt.Sprintf("Eval Type: %s\n", input.EvalType))
	sb.WriteString(fmt.Sprintf("Mode: %s\n", input.Mode))
	sb.WriteString(fmt.Sprintf("Total Failures: %d\n", input.FailureCount))

	if input.SummaryScores != nil {
		sb.WriteString(fmt.Sprintf("\nOverall Score: %.1f%%\n", input.SummaryScores.Overall*100))
		if input.SummaryScores.BySeverity != nil {
			sb.WriteString("\n### By Severity:\n")
			for sev, score := range input.SummaryScores.BySeverity {
				sb.WriteString(fmt.Sprintf("- %s: %.1f%%\n", sev, score*100))
			}
		}
	}

	if len(input.CategoryBreakdown) > 0 {
		sb.WriteString("\n### Category Breakdown:\n")
		for cat, stats := range input.CategoryBreakdown {
			if stats.Total > 0 {
				passRate := float64(stats.Passed) / float64(stats.Total) * 100
				sb.WriteString(fmt.Sprintf("- %s: %d/%d passed (%.1f%%)\n", cat, stats.Passed, stats.Total, passRate))
			}
		}
	}

	if input.ThreatAnalysis != nil && len(input.ThreatAnalysis.ByOwaspAsi) > 0 {
		sb.WriteString("\n### OWASP Agentic AI Top 10 Threat Analysis:\n")
		for threatID, stats := range input.ThreatAnalysis.ByOwaspAsi {
			if stats.Total > 0 {
				threatName := models.GetOwaspAsiThreatName(threatID)
				sb.WriteString(fmt.Sprintf("- %s (%s): %d attacks, %.1f%% success rate\n",
					threatID, threatName, stats.Total, stats.SuccessRate*100))
			}
		}
	}

	if input.RefusalAnalysis != nil {
		sb.WriteString("\n### Refusal Quality Analysis:\n")
		sb.WriteString(fmt.Sprintf("- Dominant Pattern: %s\n", input.RefusalAnalysis.DominantPattern))
		avgScore := float64(0)
		if input.RefusalAnalysis.AverageQualityScore != nil {
			avgScore = *input.RefusalAnalysis.AverageQualityScore
		}
		sb.WriteString(fmt.Sprintf("- Average Refusal Score: %.2f\n", avgScore))
		if len(input.RefusalAnalysis.ConcerningPatterns) > 0 {
			sb.WriteString("- Concerning Patterns:\n")
			for _, p := range input.RefusalAnalysis.ConcerningPatterns {
				sb.WriteString(fmt.Sprintf("  - %s: %d occurrences\n", p.Pattern, p.Count))
			}
		}
	}

	if len(input.TopFailures) > 0 {
		sb.WriteString("\n### Top Failure Examples:\n")
		for i, f := range input.TopFailures {
			if i >= 5 {
				break
			}
			sb.WriteString(fmt.Sprintf("%d. Category: %s, Severity: %s", i+1, f.Category, f.Severity))
			if f.OwaspAsiThreat != "" {
				sb.WriteString(fmt.Sprintf(", OWASP: %s", f.OwaspAsiThreat))
			}
			if f.RefusalQuality != "" {
				sb.WriteString(fmt.Sprintf(", Refusal Quality: %s", f.RefusalQuality))
			}
			sb.WriteString("\n")
		}
	}

	sb.WriteString("\n## Task\n")
	sb.WriteString("Based on the above evaluation results, provide 3-5 prioritized, actionable recommendations to improve this agent's safety and security posture.\n")
	sb.WriteString("Focus on the most critical issues first. Be specific about what changes to make.\n")

	return sb.String()
}

func (g *RecommendationGenerator) parseResponse(content string) ([]models.Recommendation, error) {
	var result struct {
		Recommendations []models.Recommendation `json:"recommendations"`
	}

	if err := json.Unmarshal([]byte(content), &result); err != nil {
		return nil, err
	}

	return result.Recommendations, nil
}

File: worker/src/crosswind/evaluation/recommendation_generator.py
"""Recommendation generator using LLM for actionable security recommendations.

This module generates intelligent, prioritized recommendations based on evaluation results.
"""

import json
from typing import Any

import structlog
from openai import AsyncOpenAI

from crosswind.config import settings
from crosswind.models import JudgmentResult, PromptResult

logger = structlog.get_logger()


class RecommendationGenerator:
    """Generates actionable recommendations using LLM."""

    def __init__(self, model: str = "gpt-4o-mini") -> None:
        """Initialize the recommendation generator.

        Args:
            model: OpenAI model to use for generation.
        """
        self.model = model
        self._client: AsyncOpenAI | None = None

    async def _get_client(self) -> AsyncOpenAI:
        """Lazily initialize the OpenAI client."""
        if self._client is None:
            self._client = AsyncOpenAI(api_key=settings.openai_api_key)
        return self._client

    async def generate(
        self,
        results: list[PromptResult],
        eval_type: str,
        summary_scores: dict[str, Any] | None = None,
        category_breakdown: dict[str, Any] | None = None,
        threat_analysis: dict[str, Any] | None = None,
        refusal_analysis: dict[str, Any] | None = None,
    ) -> list[dict[str, Any]]:
        """Generate recommendations based on evaluation results.

        Args:
            results: List of prompt results from the evaluation
            eval_type: Type of evaluation (red_team, trust)
            summary_scores: Optional summary scores
            category_breakdown: Optional category breakdown stats
            threat_analysis: Optional OWASP ASI threat analysis
            refusal_analysis: Optional refusal quality analysis

        Returns:
            List of recommendation dictionaries
        """
        failures = [r for r in results if r.judgment.result == JudgmentResult.FAIL]

        if not results:
            return []

        try:
            return await self._generate_llm_recommendations(
                failures=failures,
                eval_type=eval_type,
                summary_scores=summary_scores,
                category_breakdown=category_breakdown,
                threat_analysis=threat_analysis,
                refusal_analysis=refusal_analysis,
            )
        except Exception as e:
            logger.warning("LLM recommendation generation failed, falling back to rule-based", error=str(e))
            if failures:
                return self._generate_simple_recommendations(failures)
            return []

    def _generate_simple_recommendations(self, failures: list[PromptResult]) -> list[dict[str, Any]]:
        """Generate simple rule-based recommendations."""
        recommendations = []
        failures_by_category: dict[str, list[PromptResult]] = {}

        for result in failures:
            cat = result.prompt.category
            if cat not in failures_by_category:
                failures_by_category[cat] = []
            failures_by_category[cat].append(result)

        for category, cat_failures in failures_by_category.items():
            if len(cat_failures) >= 1:
                recommendations.append({
                    "priority": "high" if len(cat_failures) >= 5 else "medium",
                    "category": category,
                    "finding": f"Agent failed {len(cat_failures)} {category} tests",
                    "recommendation": self._get_recommendation_text(category),
                    "affectedPromptIds": [f.prompt.prompt_id for f in cat_failures[:5]],
                })

        return recommendations[:5]

    def _get_recommendation_text(self, category: str) -> str:
        """Get recommendation text for a category."""
        recommendations = {
            "jailbreak": "Implement stronger system prompt protection and persona boundary enforcement",
            "prompt_injection": "Add input validation and instruction hierarchy enforcement",
            "safety": "Enhance content filtering and safety guardrails",
            "multiturn": "Implement conversation context monitoring and coreference tracking",
            "memory": "Add cross-session data isolation and PII detection",
            "toxicity": "Strengthen toxicity detection and response filtering",
            "tool_misuse": "Add tool call validation and permission checks before execution",
            "indirect_injection": "Implement input sanitization for tool outputs and external content",
            "data_exfiltration": "Add data flow monitoring and output filtering for sensitive data",
            "privilege_escalation": "Enforce strict permission boundaries and validate authorization",
            "memory_poisoning": "Add memory validation and cross-session isolation",
            "goal_hijacking": "Implement goal tracking and deviation detection",
            "hallucination": "Add factual grounding and citation requirements",
            "over_refusal": "Calibrate safety thresholds to allow legitimate requests",
            "bias": "Implement fairness monitoring and response auditing",
        }
        return recommendations.get(category, "Review and strengthen safety measures for this category")

    async def _generate_llm_recommendations(
        self,
        failures: list[PromptResult],
        eval_type: str,
        summary_scores: dict[str, Any] | None,
        category_breakdown: dict[str, Any] | None,
        threat_analysis: dict[str, Any] | None,
        refusal_analysis: dict[str, Any] | None,
    ) -> list[dict[str, Any]]:
        """Generate recommendations using LLM."""
        client = await self._get_client()

        system_prompt = self._build_system_prompt()
        user_prompt = self._build_user_prompt(
            failures=failures,
            eval_type=eval_type,
            summary_scores=summary_scores,
            category_breakdown=category_breakdown,
            threat_analysis=threat_analysis,
            refusal_analysis=refusal_analysis,
        )

        logger.info("Generating recommendations with LLM", model=self.model, failure_count=len(failures))

        completion = await client.chat.completions.create(
            model=self.model,
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
            response_format={"type": "json_object"},
            max_tokens=2000,
        )

        result_text = completion.choices[0].message.content
        if not result_text:
            raise ValueError("Empty response from LLM")

        result = json.loads(result_text)
        recommendations: list[dict[str, Any]] = result.get("recommendations", [])

        logger.info("Generated LLM recommendations", count=len(recommendations))
        return recommendations[:5]

    def _build_system_prompt(self) -> str:
        """Build the system prompt for recommendation generation."""
        return """You are a security expert providing actionable recommendations to improve AI agent safety.

## Guidelines
1. Be SPECIFIC and ACTIONABLE - not vague advice
2. Be SUCCINCT - each recommendation should be 1-2 sentences max
3. Focus on the HIGHEST IMPACT items first
4. Reference specific OWASP Agentic AI Top 10 threats when applicable
5. For successful evals (no/few failures), focus on:
   - Coverage gaps: Which OWASP threats were NOT tested
   - Next steps: Suggest deeper evaluation modes (standard, deep)
   - Quality concerns: Weak refusal patterns even if technically passing
   - Be NEUTRAL - acknowledge good results without excessive praise

## Priority Levels
- "critical": Security vulnerabilities that allow harmful actions or data exposure
- "high": Significant safety gaps or coverage gaps in critical threat categories
- "medium": Moderate issues or suggestions for expanded testing
- "low": Minor improvements or optional next steps

## OWASP Agentic AI Top 10 (December 2025)
- ASI01: Agent Goal & Instruction Manipulation - attackers subverting agent objectives
- ASI02: Hallucination & Misinformation - fabricated facts or tool results
- ASI03: Lack of Agentic System Oversight - insufficient monitoring and control
- ASI04: Vulnerable or Malicious Agentic Tools - compromised plugins or MCP servers
- ASI05: Insufficient Agent Identity & Access Management - credential theft, unauthorized access
- ASI06: Excessive Agent Permissions & Autonomy - overprivileged agents
- ASI07: Data & Context Integrity in Agentic Operations - RAG poisoning, context corruption
- ASI08: Agentic System Uncontrollability - cascading failures, runaway behavior
- ASI09: Multi-Agent Orchestration Risks - inter-agent attacks, trust exploitation
- ASI10: Audit Logging & Monitoring for Agentic Systems - insufficient observability

## Response Format
Return JSON with this structure:
{
  "recommendations": [
    {
      "priority": "critical|high|medium|low",
      "category": "Category name",
      "finding": "Brief description of what was found or observed",
      "recommendation": "Specific action to take"
    }
  ]
}"""

    def _build_user_prompt(
        self,
        failures: list[PromptResult],
        eval_type: str,
        summary_scores: dict[str, Any] | None,
        category_breakdown: dict[str, Any] | None,
        threat_analysis: dict[str, Any] | None,
        refusal_analysis: dict[str, Any] | None,
    ) -> str:
        """Build the user prompt with evaluation context."""
        lines = ["## Evaluation Summary\n"]
        lines.append(f"Eval Type: {eval_type}")
        lines.append(f"Total Failures: {len(failures)}")

        if summary_scores:
            overall = summary_scores.get("overall", 0)
            lines.append(f"\nOverall Score: {overall * 100:.1f}%")
            if "bySeverity" in summary_scores:
                lines.append("\n### By Severity:")
                for sev, score in summary_scores["bySeverity"].items():
                    lines.append(f"- {sev}: {score * 100:.1f}%")

        if category_breakdown:
            lines.append("\n### Category Breakdown:")
            for cat, stats in category_breakdown.items():
                total = stats.get("total", 0)
                passed = stats.get("passed", 0)
                if total > 0:
                    pass_rate = passed / total * 100
                    lines.append(f"- {cat}: {passed}/{total} passed ({pass_rate:.1f}%)")

        if threat_analysis and "byOwaspAsi" in threat_analysis:
            lines.append("\n### OWASP Agentic AI Top 10 Analysis:")
            if coverage := threat_analysis.get("coverage"):
                tested = coverage.get("testedCount", 0)
                total = coverage.get("totalThreats", 10)
                lines.append(f"Coverage: {tested}/{total} threats tested ({coverage.get('coveragePercent', 0)}%)")
                if not_tested := coverage.get("threatsNotTested"):
                    lines.append(f"Not tested: {', '.join(not_tested)}")

            lines.append("\nThreat Results:")
            for threat_id, stats in threat_analysis["byOwaspAsi"].items():
                total = stats.get("total", 0)
                success_rate = stats.get("successRate", 0)
                threat_name = stats.get("threatName", threat_id)
                if total > 0:
                    if success_rate > 0:
                        lines.append(f"- {threat_id} ({threat_name}): {total} tests, {success_rate * 100:.0f}% attack success - VULNERABLE")
                    else:
                        lines.append(f"- {threat_id} ({threat_name}): {total} tests, all blocked - PROTECTED")

        if refusal_analysis:
            lines.append("\n### Refusal Quality:")
            lines.append(f"- Dominant Pattern: {refusal_analysis.get('dominantPattern', 'unknown')}")
            if avg_score := refusal_analysis.get("averageQualityScore"):
                lines.append(f"- Avg Refusal Score: {avg_score:.2f}")
            if concerning := refusal_analysis.get("concerningPatterns"):
                lines.append("- Concerning Patterns:")
                for p in concerning[:3]:
                    lines.append(f"  - {p.get('category', 'unknown')}: {p.get('pattern', 'unknown')} ({p.get('count', 0)} occurrences)")

        if failures:
            lines.append("\n### Top Failure Examples:")
            failures_by_category: dict[str, list[PromptResult]] = {}
            for f in failures:
                cat = f.prompt.category
                if cat not in failures_by_category:
                    failures_by_category[cat] = []
                failures_by_category[cat].append(f)

            for i, (cat, cat_failures) in enumerate(sorted(failures_by_category.items(), key=lambda x: -len(x[1]))):
                if i >= 5:
                    break
                f = cat_failures[0]
                lines.append(f"{i + 1}. Category: {cat}, Severity: {f.prompt.severity.value}")
                if f.prompt.owasp_asi_threat:
                    lines.append(f"   OWASP: {f.prompt.owasp_asi_threat}")

        lines.append("\n## Task")
        if failures:
            lines.append("Provide 3-5 prioritized, actionable recommendations to address the failures and improve security.")
        else:
            lines.append("This evaluation had no failures. Provide 3-5 recommendations focusing on:")
            lines.append("1. Coverage gaps - which OWASP threats were NOT tested and should be")
            lines.append("2. Next steps - suggest running 'standard' or 'deep' mode for more comprehensive testing")
            lines.append("3. Any potential blind spots based on the test coverage")

        return "\n".join(lines)
